app/models/manageiq/providers/storage_manager/cinder_manager/refresh_parser/cross_linkers/openstack.rb

```diff
@@ -0,0 +1,117 @@
+module ManageIQ::Providers::StorageManager::CinderManager::RefreshParser::CrossLinkers
+  class Openstack
+    include Vmdb::Logging
+
+    def initialize(parent_ems, data)
+      @parent_ems = parent_ems
+      @data       = data
+
+      @parent_ems.cloud_tenants.reset
+      @parent_ems.availability_zones.reset
+      @parent_ems.vms.reset
+    end
+
+    def cross_link
+      @data[:cloud_volumes].each do |volume_hash|
+        api_obj = volume_hash[:api_obj]
+
+        link_volume_to_tenant(volume_hash, api_obj)
+        link_volume_to_availability_zone(volume_hash, api_obj)
+        link_volume_to_disk(volume_hash, api_obj)
+      end
+
+      @data[:cloud_volume_snapshots].each do |snapshot_hash|
+        api_obj = snapshot_hash[:api_obj]
+
+        link_snapshot_to_tenant(snapshot_hash, api_obj)
+      end
+    end
+
+    def link_volume_to_tenant(volume_hash, api_obj)
+      tenant = @parent_ems.cloud_tenants.try(:where, :ems_ref => api_obj.tenant_id).try(:first)
+      unless tenant
+        _log.info "EMS: #{@parent_ems.name}, tenant not found: #{api_obj.tenant_id}"
+        return
+      end
+      _log.debug "Found tenant: #{api_obj.tenant_id}, id = #{tenant.id}"
+
+      volume_hash[:cloud_tenant_id] = tenant.id
+    end
+
+    def link_volume_to_availability_zone(volume_hash, api_obj)
+      az_ref = "volume-#{api_obj.availability_zone}"
+      availability_zone = @parent_ems.availability_zones.try(:where, :ems_ref => az_ref).try(:first)
```

Comment:
Prefer iterating over the list in memory instead of issuing a new query on each pass.  (and in the other places in this file)